Use versioned sourceset for REST compat testing

jakelandis opened the pull request:
This commit changes the sourceset name for REST API compatility testing 
from yamlRestCompatTest to yamlRestTestV(N)Compat wher `(N)` is the version
from which the tests are sourced. For example yamlRestTestV7Compat 
or yamlRestTestV8Compat.

Two motivating reasons for this change. 

1) consitent naming - all compatible rest testing will now use a common
prefix  of `yamlRestTestV7Compat` where the task name to execute is 
`yamlRestTestV7CompatTest`and the transformation task is `yamlRestTestV7CompatTransform`
It makes sense that corresponding source set follows this convention.

2) Better support when the major version changes. Gradle uses the source
set under the covers to execute the compatiblity tests. However, a user 
can add custom REST tests to this source set. By using dynamically named 
source sets, there should little concern of accidentally running the wrong
version tests. Additionaly it is visualy easy to tell which version the tests are 
from. 

This commit also removes the unecessary duplicated v7compat directory
and adds copy and classpath duplicate checks to ensure that we don't 
accidentally stomp on duplicates with all of the copy and classpath manipulation.  

related: https://github.com/elastic/elasticsearch/pull/75267

jakelandis committed:
Use versioned sourceset for REST compat testing

jakelandis committed:
rename

jakelandis committed:
fix typo

jakelandis committed:
add Compat to sourceset name

jakelandis committed:
Merge remote-tracking branch 'upstream/master' into use_version_source_set

jakelandis committed:
minor fixes

jakelandis committed:
update dir names

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/rest/compat/YamlRestCompatTestPlugin.java:
```diff
@@ -55,6 +54,7 @@
     public void apply(Project project) {
         final int compatibleVersion = Version.fromString(VersionProperties.getVersions().get("elasticsearch")).getMajor() - 1;
+        final String sourceSetName = "yamlRestTestV" + compatibleVersion + "Compat";
         final Path compatRestResourcesDir = Path.of("restResources").resolve("v" + compatibleVersion);
```
If we are having version-specific source sets, do we still need to capture the version here in the file path as well. As it is now we have files like `src/yamlRestTestV7Compat/resources/rest-api-spec/test/v7compat/foo/bar.yml`. We probably don't need to track the version twice now. Also begs the question whether we need this change in the first place if we are _already_ managing per-version test resources. How will this affect backporting, etc?

jakelandis replied:
`src/yamlRestTestV7Compat/resources/rest-api-spec/test/v7compat/foo/bar.yml` uses a 4 part path to keept it from colliding with the actual v7 `test/foo/bar.yml` if it exists. (Both the real v7 tests and the custom test in this source set are on the same classpath, and there quite a bit of re-use of file names).  `v7compat` in the path is just a convention and included the v7 to help demark these are intended for compatibilty with v7. The fact this is just a convention and without any teeth (i.e. these would execute when master is v9) is a motivating factor for this PR. 

I can remove the `v7` from the `v7compat` convention in this PR since I agree it doesn't make sense now. I think I could also remove this extra directory entirely and ensure uniqueness of the files between the two sources of tests (with validation in Gradle). Thoughts ?

Since our REST compatilbity only supports 1 major version, these should never be backported and anything in this sourceset should be removed in the next major version.

mark-vieira replied:
> I can remove the v7 from the v7compat convention in this PR since I agree it doesn't make sense now. I think I could also remove this extra directory entirely and ensure uniqueness of the files between the two sources of tests (with validation in Gradle). Thoughts ?

Yeah, I think that makes sense. It's just less stuff to try and keep in sync.

jakelandis commented on x-pack/plugin/build.gradle:
```diff
@@ -18,7 +18,7 @@ dependencies {
 // let the yamlRestTests see the classpath of test
 GradleUtils.extendSourceSet(project, "test", "yamlRestTest", tasks.named("yamlRestTest"))
-GradleUtils.extendSourceSet(project, "test", "yamlRestCompatTest")
+GradleUtils.extendSourceSet(project, "test", "yamlRestTestV7Compat") //TODO: use version resolution
```
@mark-vieira - Any suggestions for how read the current version so I don't need to hard code the version here ?

mark-vieira replied:
Just use `VersionProperties.getElasticsearchVersion().getMajor()` to get the current major. Then I assume we just decrement it.

jakelandis replied:
perfect .. thanks!

jakelandis committed:
fix todo

jakelandis committed:
implement duplicate check and move stuff around

jakelandis committed:
moved stuff around

jakelandis committed:
Merge branch 'master' of github.com:elastic/elasticsearch into use_version_source_set

jakelandis commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/CopyRestApiTask.java:
```diff
@@ -126,13 +127,15 @@ void copy() {
 
         getLogger().debug("Rest specs for project [{}] will be copied to the test resources.", projectPath);
         fileSystemOperations.copy(c -> {
+            c.setDuplicatesStrategy(DuplicatesStrategy.FAIL);
```
These duplicate fails are not actually needed for this PR, but I realized while adding the custom duplicate check to the compat testing task, it would be good to add a duplicate check here too.

mark-vieira replied:
I don't think this is necessary since `FAIL` is now the default behavior in Gradle 7.0 and later.

jakelandis replied:
ah.. it is. https://github.com/gradle/gradle/pull/15759  I will remove this.

jakelandis commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/rest/compat/YamlRestCompatTestPlugin.java:
```diff
@@ -178,11 +184,33 @@ public void apply(Project project) {
+                            + ".x] branch and the [" + sourceSetName + "] tests. ");
+                    }
+                }
+            });
```
example of failure:
```
* What went wrong:
Execution failed for task ':rest-api-spec:yamlRestTestV7CompatTest'.
> Found duplicated test(s) [search/30_limits.yml] please ensure there not any duplicate YAML test files between the [7.x] branch and the [yamlRestTestV7Compat] tests. 
```

jakelandis commented:
@mark-vieira - should be ready for another review.  I implemented a duplicate check to ensure that the same test file is not on the class path more then once. 98% of the tests use 2 parts (foo/bar.yml) under the folder test and the duplicate check checks for those 2 parts. _There is a slim chance of a false positive if a test uses 3 parts and the custom compat uses 2 part and there is name collision, but that is extremely unlikely._

I also removed the uncessary `v7compat` directory now that there is a proper check for duplicates.

jakelandis committed:
remove unecessary DuplicatesStrategy.FAIL

jakelandis committed:
fix import

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into use_version_source_set

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

jakelandis committed:
Merge remote-tracking branch 'upstream/master' into use_version_source_set

jakelandis committed:
fix merge issue

jakelandis committed:
remove gradle duplicate check in favor of testing framwork fix

jakelandis commented:
> I'm wondering if it might be better (and simpler) to implement this in the yaml testing framework code instead. It could be as simple as checking the result of the call to add() here and throwing an exception if it returns false.

Changed on [`2a8fc87` (#78418)](https://github.com/elastic/elasticsearch/pull/78418/commits/2a8fc873657d78593e88fd97c40dbf017bf49a6c)

It turns out we actually do support running duplicates on the classpath without stomping on each other. The framework will append a #2 to each test name. In light of this, i reduced the error down to a warning since which test gets the #2 appended is classloader order. We probably could add in some specific ordering of the collections so it wouldn't be classloader ordering but it would be some other non obvious ordering. The new warning is there in part to inform future developers that we actually do support duplicated names. 

Example of the new warning message:
```
org.elasticsearch.test.rest.ClientYamlTestSuiteIT STANDARD_OUT
    [2021-10-11T19:44:38,076][WARN ][o.e.t.r.y.ESClientYamlSuiteTestCase] [[Test worker]] Found duplicate test name [search/10_source_filtering.yml] on the class path. This can result in class loader dependent execution commands and reproduction commands (will add #2 to one of the test names dependent on the classloading order)
```

I don't have a strong opinion to keep or leave this warning.

jakelandis committed:
static and import cleanup

jakelandis commented:
@mark-vieira - finally got back round to this. should be ready for another review

mark-vieira reviewed (approved):
LGTM

elasticmachine committed:
Merge branch 'master' into use_version_source_set
